From b5c269bbe54cf4555d883c2a5d145503fadb5b8e Mon Sep 17 00:00:00 2001
From: T S Rameshkumar <rameshkumar.t@phytecembedded.in>
Date: Fri, 9 Oct 2026 18:41:51 +0530
Subject: [PATCH 01/13] imx8m: phyboard_polis: Add target build makefiles and
 board configs Add phyboard_polis.mk, BoardConfig.mk, AndroidProducts.mk lunch
 choice (phyboard_polis-userdebug), and U-Boot build script for phyBOARD-Polis
 i.MX8MM.

---
 imx8m/AndroidProducts.mk                      |   7 +-
 imx8m/phyboard_polis/AndroidBoard.mk          |  10 +
 imx8m/phyboard_polis/AndroidUboot.sh          |  90 +++
 imx8m/phyboard_polis/BoardConfig.mk           | 185 +++++++
 imx8m/phyboard_polis/SharedBoardConfig.mk     | 191 +++++++
 .../phyboard_polis/UbootKernelBoardConfig.mk  |  21 +
 imx8m/phyboard_polis/phyboard_polis.mk        | 520 ++++++++++++++++++
 7 files changed, 1022 insertions(+), 2 deletions(-)
 create mode 100644 imx8m/phyboard_polis/AndroidBoard.mk
 create mode 100644 imx8m/phyboard_polis/AndroidUboot.sh
 create mode 100644 imx8m/phyboard_polis/BoardConfig.mk
 create mode 100644 imx8m/phyboard_polis/SharedBoardConfig.mk
 create mode 100644 imx8m/phyboard_polis/UbootKernelBoardConfig.mk
 create mode 100644 imx8m/phyboard_polis/phyboard_polis.mk

diff --git a/imx8m/AndroidProducts.mk b/imx8m/AndroidProducts.mk
index 59f449f8..169fe20a 100644
--- a/imx8m/AndroidProducts.mk
+++ b/imx8m/AndroidProducts.mk
@@ -30,7 +30,8 @@ PRODUCT_MAKEFILES := \
   $(LOCAL_DIR)/evk_8mq/evk_8mq.mk \
   $(LOCAL_DIR)/evk_8mm/evk_8mm.mk \
   $(LOCAL_DIR)/evk_8mn/evk_8mn.mk \
-  $(LOCAL_DIR)/evk_8mp/evk_8mp.mk
+  $(LOCAL_DIR)/evk_8mp/evk_8mp.mk \
+  $(LOCAL_DIR)/phyboard_polis/phyboard_polis.mk
 
 COMMON_LUNCH_CHOICES := \
   evk_8mq-nxp_stable-user \
@@ -40,4 +41,6 @@ COMMON_LUNCH_CHOICES := \
   evk_8mn-nxp_stable-user \
   evk_8mn-nxp_stable-userdebug \
   evk_8mp-nxp_stable-user \
-  evk_8mp-nxp_stable-userdebug
+  evk_8mp-nxp_stable-userdebug \
+  phyboard_polis-user \
+  phyboard_polis-userdebug
diff --git a/imx8m/phyboard_polis/AndroidBoard.mk b/imx8m/phyboard_polis/AndroidBoard.mk
new file mode 100644
index 00000000..e658dbd0
--- /dev/null
+++ b/imx8m/phyboard_polis/AndroidBoard.mk
@@ -0,0 +1,10 @@
+LOCAL_PATH := $(call my-dir)
+
+include $(CONFIG_REPO_PATH)/common/build/dtbo.mk
+include $(CONFIG_REPO_PATH)/common/build/imx-recovery.mk
+include $(CONFIG_REPO_PATH)/common/build/gpt.mk
+include $(FSL_PROPRIETARY_PATH)/fsl-proprietary/media-profile/media-profile.mk
+-include $(IMX_MEDIA_CODEC_XML_PATH)/mediacodec-profile/mediacodec-profile.mk
+
+BOARD_PACK_RADIOIMAGES += bootloader.img
+INSTALLED_RADIOIMAGE_TARGET  += $(PRODUCT_OUT)/bootloader.img
diff --git a/imx8m/phyboard_polis/AndroidUboot.sh b/imx8m/phyboard_polis/AndroidUboot.sh
new file mode 100644
index 00000000..74f9bbcc
--- /dev/null
+++ b/imx8m/phyboard_polis/AndroidUboot.sh
@@ -0,0 +1,90 @@
+#!/bin/bash
+
+# hardcode this one again in this shell script
+CONFIG_REPO_PATH=device/nxp
+
+# import other paths in the file "common/imx_path/ImxPathConfig.mk" of this
+# repository
+
+while read -r line
+do
+	if [ "$(echo ${line} | grep "=")" != "" ]; then
+		env_arg=`echo ${line} | cut -d "=" -f1`
+		env_arg=${env_arg%:}
+		env_arg=`eval echo ${env_arg}`
+
+		env_arg_value=`echo ${line} | cut -d "=" -f2`
+		env_arg_value=`eval echo ${env_arg_value}`
+
+		eval ${env_arg}=${env_arg_value}
+	fi
+done < ${CONFIG_REPO_PATH}/common/imx_path/ImxPathConfig.mk
+
+
+if [ "${AARCH64_GCC_CROSS_COMPILE}" != "" ]; then
+    ATF_CROSS_COMPILE=`eval echo ${AARCH64_GCC_CROSS_COMPILE}`
+else
+    echo ERROR: \*\*\* env AARCH64_GCC_CROSS_COMPILE is not set
+    exit 1
+fi
+
+build_pre_image()
+{
+	:
+}
+
+build_imx_uboot()
+{
+	echo Building i.MX U-Boot with firmware
+	export dtbs=imx8mm-phyboard-polis-rdk.dtb
+	cp ${UBOOT_OUT}/u-boot-nodtb.$1 ${IMX_MKIMAGE_PATH}/imx-mkimage/iMX8M/
+	cp ${UBOOT_OUT}/spl/u-boot-spl.bin  ${IMX_MKIMAGE_PATH}/imx-mkimage/iMX8M/
+	cp ${UBOOT_OUT}/tools/mkimage  ${IMX_MKIMAGE_PATH}/imx-mkimage/iMX8M/mkimage_uboot
+
+	cp ${UBOOT_OUT}/arch/arm/dts/imx8mm-phyboard-polis-rdk.dtb  ${IMX_MKIMAGE_PATH}/imx-mkimage/iMX8M/
+	cp ${FSL_PROPRIETARY_PATH}/linux-firmware-imx/firmware/ddr/synopsys/lpddr4_pmu_train* ${IMX_MKIMAGE_PATH}/imx-mkimage/iMX8M/
+
+
+	make -C ${IMX_PATH}/arm-trusted-firmware/ PLAT=`echo $2 | cut -d '-' -f1` clean
+	if [ "`echo $2 | cut -d '-' -f2`" = "trusty" ] && [ "`echo $2 | rev | cut -d '-' -f1`" != "uuu" ]; then
+		cp ${FSL_PROPRIETARY_PATH}/fsl-proprietary/uboot-firmware/imx8m/tee-imx8mm.bin ${IMX_MKIMAGE_PATH}/imx-mkimage/iMX8M/tee.bin
+		if [ "`echo $2 | cut -d '-' -f3`" = "4g" ]; then
+			make -C ${IMX_PATH}/arm-trusted-firmware/ CROSS_COMPILE="${ATF_CROSS_COMPILE}" PLAT=`echo $2 | cut -d '-' -f1` bl31 -B BL32_BASE=0xfe000000 SPD=trusty IMX_ANDROID_BUILD=true IMX_BOOT_UART_BASE=0x30880000 1>/dev/null || exit 1
+		else
+			make -C ${IMX_PATH}/arm-trusted-firmware/ CROSS_COMPILE="${ATF_CROSS_COMPILE}" PLAT=`echo $2 | cut -d '-' -f1` bl31 -B SPD=trusty IMX_ANDROID_BUILD=true IMX_BOOT_UART_BASE=0x30880000 1>/dev/null || exit 1
+		fi
+	else
+		if [ -f ${IMX_MKIMAGE_PATH}/imx-mkimage/iMX8M/tee.bin ] ; then
+			rm -rf ${IMX_MKIMAGE_PATH}/imx-mkimage/iMX8M/tee.bin
+		fi
+		if [ -f ${IMX_MKIMAGE_PATH}/imx-mkimage/iMX8M/tee.bin.lz4 ] ; then
+			rm -rf ${IMX_MKIMAGE_PATH}/imx-mkimage/iMX8M/tee.bin.lz4
+		fi
+		make -C ${IMX_PATH}/arm-trusted-firmware/ CROSS_COMPILE="${ATF_CROSS_COMPILE}" PLAT=`echo $2 | cut -d '-' -f1` bl31 -B IMX_ANDROID_BUILD=true IMX_BOOT_UART_BASE=0x30880000 1>/dev/null || exit 1
+	fi
+	cp ${IMX_PATH}/arm-trusted-firmware/build/`echo $2 | cut -d '-' -f1`/release/bl31.bin ${IMX_MKIMAGE_PATH}/imx-mkimage/iMX8M/bl31.bin
+
+
+	make -C ${IMX_MKIMAGE_PATH}/imx-mkimage/ clean
+	if [ `echo $2 | cut -d '-' -f2` = "ddr4" ]; then
+		make -C ${IMX_MKIMAGE_PATH}/imx-mkimage/ SOC=iMX8MM dtbs=imx8mm-phyboard-polis-rdk.dtb flash_ddr4_evk || exit 1
+		make -C ${IMX_MKIMAGE_PATH}/imx-mkimage/ SOC=iMX8MM dtbs=imx8mm-phyboard-polis-rdk.dtb print_fit_hab_ddr4 || exit 1
+	elif [ `echo $2 | cut -d '-' -f2` = "4g" ] || [ "`echo $2 | cut -d '-' -f3`" = "4g" ] || [ `echo $2 | rev | cut -d '-' -f1` = "uuu" ]; then
+		make -C ${IMX_MKIMAGE_PATH}/imx-mkimage/ SOC=iMX8MM dtbs=imx8mm-phyboard-polis-rdk.dtb TEE_LOAD_ADDR=0xfe000000 flash_spl_uboot || exit 1
+		make -C ${IMX_MKIMAGE_PATH}/imx-mkimage/ SOC=iMX8MM dtbs=imx8mm-phyboard-polis-rdk.dtb print_fit_hab || exit 1
+	elif [ `echo $2 | rev | cut -d '-' -f1 | rev` != "dual" ]; then
+		make -C ${IMX_MKIMAGE_PATH}/imx-mkimage/ SOC=iMX8MM dtbs=imx8mm-phyboard-polis-rdk.dtb flash_spl_uboot || exit 1
+		make -C ${IMX_MKIMAGE_PATH}/imx-mkimage/ SOC=iMX8MM dtbs=imx8mm-phyboard-polis-rdk.dtb print_fit_hab || exit 1
+	else
+		make -C ${IMX_MKIMAGE_PATH}/imx-mkimage/ SOC=iMX8MM dtbs=imx8mm-phyboard-polis-rdk.dtb flash_evk_no_hdmi_dual_bootloader || exit 1
+		make -C ${IMX_MKIMAGE_PATH}/imx-mkimage/ SOC=iMX8MM dtbs=imx8mm-phyboard-polis-rdk.dtb PRINT_FIT_HAB_OFFSET=0x0 print_fit_hab || exit 1
+	fi
+
+
+	if [ `echo $2 | rev | cut -d '-' -f1 | rev` != "dual" ]; then
+		cp ${IMX_MKIMAGE_PATH}/imx-mkimage/iMX8M/flash.bin ${UBOOT_COLLECTION}/u-boot-$2.imx
+	else
+		cp ${IMX_MKIMAGE_PATH}/imx-mkimage/iMX8M/flash.bin ${UBOOT_COLLECTION}/spl-$2.bin
+		cp ${IMX_MKIMAGE_PATH}/imx-mkimage/iMX8M/u-boot-ivt.itb ${UBOOT_COLLECTION}/bootloader-$2.img
+	fi
+}
diff --git a/imx8m/phyboard_polis/BoardConfig.mk b/imx8m/phyboard_polis/BoardConfig.mk
new file mode 100644
index 00000000..718c147e
--- /dev/null
+++ b/imx8m/phyboard_polis/BoardConfig.mk
@@ -0,0 +1,185 @@
+# -------@block_infrastructure-------
+#
+# Product-specific compile-time definitions.
+#
+
+include $(CONFIG_REPO_PATH)/imx8m/BoardConfigCommon.mk
+
+# -------@block_common_config-------
+#
+# SoC-specific compile-time definitions.
+#
+
+# value assigned in this part should be fixed for an SoC, right?
+
+BOARD_SOC_TYPE := IMX8MM
+BOARD_HAVE_VPU := true
+BOARD_VPU_TYPE := hantro
+FSL_VPU_OMX_ONLY := true
+HAVE_FSL_IMX_GPU2D := true
+HAVE_FSL_IMX_GPU3D := true
+HAVE_FSL_IMX_PXP := false
+TARGET_USES_HWC2 := true
+TARGET_HAVE_VULKAN := false
+
+SOONG_CONFIG_IMXPLUGIN += \
+                          BOARD_VPU_TYPE
+
+SOONG_CONFIG_IMXPLUGIN_BOARD_SOC_TYPE = IMX8MM
+SOONG_CONFIG_IMXPLUGIN_BOARD_HAVE_VPU = true
+SOONG_CONFIG_IMXPLUGIN_BOARD_VPU_TYPE = hantro
+SOONG_CONFIG_IMXPLUGIN_BOARD_VPU_ONLY = false
+SOONG_CONFIG_IMXPLUGIN_PREBUILT_FSL_IMX_CODEC = true
+SOONG_CONFIG_IMXPLUGIN_POWERSAVE = false
+
+# -------@block_storage-------
+TARGET_USERIMAGES_USE_EXT4 := true
+
+# use sparse image
+TARGET_USERIMAGES_SPARSE_EXT_DISABLED := false
+
+# Support gpt
+ifeq ($(TARGET_USE_DYNAMIC_PARTITIONS),true)
+  BOARD_BPT_INPUT_FILES += $(CONFIG_REPO_PATH)/common/partition/device-partitions-13GB-ab_super.bpt
+  ADDITION_BPT_PARTITION = partition-table-28GB:$(CONFIG_REPO_PATH)/common/partition/device-partitions-28GB-ab_super.bpt \
+                           partition-table-dual:$(CONFIG_REPO_PATH)/common/partition/device-partitions-13GB-ab-dual-bootloader_super.bpt \
+                           partition-table-28GB-dual:$(CONFIG_REPO_PATH)/common/partition/device-partitions-28GB-ab-dual-bootloader_super.bpt
+else
+  ifeq ($(IMX_NO_PRODUCT_PARTITION),true)
+    BOARD_BPT_INPUT_FILES += $(CONFIG_REPO_PATH)/common/partition/device-partitions-13GB-ab-no-product.bpt
+    ADDITION_BPT_PARTITION = partition-table-28GB:$(CONFIG_REPO_PATH)/common/partition/device-partitions-28GB-ab-no-product.bpt \
+                             partition-table-dual:$(CONFIG_REPO_PATH)/common/partition/device-partitions-13GB-ab-dual-bootloader-no-product.bpt \
+                             partition-table-28GB-dual:$(CONFIG_REPO_PATH)/common/partition/device-partitions-28GB-ab-dual-bootloader-no-product.bpt
+  else
+    BOARD_BPT_INPUT_FILES += $(CONFIG_REPO_PATH)/common/partition/device-partitions-13GB-ab.bpt
+    ADDITION_BPT_PARTITION = partition-table-28GB:$(CONFIG_REPO_PATH)/common/partition/device-partitions-28GB-ab.bpt \
+                             partition-table-dual:$(CONFIG_REPO_PATH)/common/partition/device-partitions-13GB-ab-dual-bootloader.bpt \
+                             partition-table-28GB-dual:$(CONFIG_REPO_PATH)/common/partition/device-partitions-28GB-ab-dual-bootloader.bpt
+  endif
+endif
+
+BOARD_PREBUILT_DTBOIMAGE := $(OUT_DIR)/target/product/$(PRODUCT_DEVICE)/dtbo-imx8mm.img
+
+BOARD_USES_METADATA_PARTITION := true
+BOARD_ROOT_EXTRA_FOLDERS += metadata
+
+ifneq ($(BUILD_ENCRYPTED_BOOT),true)
+  AB_OTA_PARTITIONS += bootloader
+endif
+
+# -------@block_security-------
+ENABLE_CFI=true
+
+BOARD_AVB_ENABLE := true
+BOARD_AVB_ALGORITHM := SHA256_RSA4096
+# The testkey_rsa4096.pem is copied from external/avb/test/data/testkey_rsa4096.pem
+BOARD_AVB_KEY_PATH := $(CONFIG_REPO_PATH)/common/security/testkey_rsa4096.pem
+
+BOARD_AVB_BOOT_KEY_PATH := external/avb/test/data/testkey_rsa4096.pem
+BOARD_AVB_BOOT_ALGORITHM := SHA256_RSA4096
+BOARD_AVB_BOOT_ROLLBACK_INDEX_LOCATION := 2
+
+# Enable chained vbmeta for init_boot images
+BOARD_AVB_INIT_BOOT_KEY_PATH := external/avb/test/data/testkey_rsa4096.pem
+BOARD_AVB_INIT_BOOT_ALGORITHM := SHA256_RSA4096
+BOARD_AVB_INIT_BOOT_ROLLBACK_INDEX_LOCATION := 3
+
+# Use sha256 hashtree
+BOARD_AVB_SYSTEM_ADD_HASHTREE_FOOTER_ARGS += --hash_algorithm sha256
+BOARD_AVB_SYSTEM_EXT_ADD_HASHTREE_FOOTER_ARGS += --hash_algorithm sha256
+BOARD_AVB_PRODUCT_ADD_HASHTREE_FOOTER_ARGS += --hash_algorithm sha256
+BOARD_AVB_VENDOR_ADD_HASHTREE_FOOTER_ARGS += --hash_algorithm sha256
+BOARD_AVB_VENDOR_DLKM_ADD_HASHTREE_FOOTER_ARGS += --hash_algorithm sha256
+BOARD_AVB_SYSTEM_DLKM_ADD_HASHTREE_FOOTER_ARGS += --hash_algorithm sha256
+
+# -------@block_treble-------
+# Vendor Interface manifest and compatibility
+DEVICE_MANIFEST_FILE := $(IMX_DEVICE_PATH)/manifest.xml
+
+DEVICE_MATRIX_FILE := $(IMX_DEVICE_PATH)/compatibility_matrix.xml
+DEVICE_FRAMEWORK_COMPATIBILITY_MATRIX_FILE := $(IMX_DEVICE_PATH)/device_framework_matrix.xml
+
+
+# -------@block_wifi-------
+# 8mm LPDDR4 board use NXP 8987 wifi
+BOARD_WLAN_DEVICE            := nxp
+WPA_SUPPLICANT_VERSION       := VER_0_8_X
+BOARD_WPA_SUPPLICANT_DRIVER  := NL80211
+BOARD_HOSTAPD_DRIVER         := NL80211
+BOARD_HOSTAPD_PRIVATE_LIB               := lib_driver_cmd_$(BOARD_WLAN_DEVICE)
+BOARD_WPA_SUPPLICANT_PRIVATE_LIB        := lib_driver_cmd_$(BOARD_WLAN_DEVICE)
+
+# NXP 8987 wifi support dual interface
+WIFI_HIDL_FEATURE_DUAL_INTERFACE := true
+
+# NXP 8987 wifi driver module
+BOARD_VENDOR_KERNEL_MODULES += \
+    $(TARGET_OUT_INTERMEDIATES)/MXMWIFI_OBJ/mlan.ko \
+    $(TARGET_OUT_INTERMEDIATES)/MXMWIFI_OBJ/moal.ko
+
+# -------@block_bluetooth-------
+# NXP 8987 bluetooth
+BOARD_HAVE_BLUETOOTH_NXP := true
+BOARD_BLUETOOTH_BDROID_BUILDCFG_INCLUDE_DIR := $(IMX_DEVICE_PATH)/bluetooth
+
+# -------@block_kernel_bootimg-------
+BOARD_KERNEL_BASE := 0x40400000
+
+CMASIZE=800M
+
+# NXP default config
+BOARD_KERNEL_CMDLINE := init=/init firmware_class.path=/vendor/firmware loop.max_part=7 bootconfig
+BOARD_BOOTCONFIG += androidboot.console=ttymxc2 androidboot.hardware=nxp
+
+# memory config
+BOARD_KERNEL_CMDLINE += transparent_hugepage=never
+BOARD_KERNEL_CMDLINE += swiotlb=512
+
+# display config
+BOARD_BOOTCONFIG += androidboot.lcd_density=240
+
+# wifi config
+BOARD_BOOTCONFIG += androidboot.wificountrycode=CN
+BOARD_KERNEL_CMDLINE += moal.mod_para=wifi_mod_para_sd8987.conf
+
+# low memory device build config
+ifeq ($(LOW_MEMORY),true)
+BOARD_KERNEL_CMDLINE += cma=320M@0x400M-0xb80M galcore.contiguousSize=33554432
+BOARD_BOOTCONFIG += androidboot.displaymode=720p
+else
+BOARD_KERNEL_CMDLINE += cma=$(CMASIZE)@0x400M-0xb80M
+endif
+
+# Disable fw_devlink.strict
+BOARD_KERNEL_CMDLINE += fw_devlink.strict=0
+
+# Add KVM support
+BOARD_BOOTCONFIG += androidboot.hypervisor.vm.supported=true
+
+ifneq (,$(filter userdebug eng,$(TARGET_BUILD_VARIANT)))
+BOARD_BOOTCONFIG += androidboot.vendor.sysrq=1
+endif
+
+ifeq ($(TARGET_USE_DYNAMIC_PARTITIONS),true)
+  ifeq ($(IMX_NO_PRODUCT_PARTITION),true)
+    TARGET_BOARD_DTS_CONFIG ?= imx8mm:imx8mm-phyboard-polis-rdk-no-product.dtb
+  else
+    TARGET_BOARD_DTS_CONFIG ?= imx8mm:imx8mm-phyboard-polis-rdk.dtb
+  endif
+else
+  ifeq ($(IMX_NO_PRODUCT_PARTITION),true)
+    TARGET_BOARD_DTS_CONFIG ?= imx8mm:imx8mm-phyboard-polis-rdk-no-product-no-dynamic_partition.dtb
+  else
+    TARGET_BOARD_DTS_CONFIG ?= imx8mm:imx8mm-phyboard-polis-rdk-no-dynamic_partition.dtb
+  endif
+endif
+
+ALL_DEFAULT_INSTALLED_MODULES += $(BOARD_VENDOR_KERNEL_MODULES)
+
+# -------@block_sepolicy-------
+BOARD_SEPOLICY_DIRS := \
+       $(CONFIG_REPO_PATH)/imx8m/sepolicy \
+       $(IMX_DEVICE_PATH)/sepolicy
+
+HAS_SYSTEM_EXT_SEPOLICY := true
+
diff --git a/imx8m/phyboard_polis/SharedBoardConfig.mk b/imx8m/phyboard_polis/SharedBoardConfig.mk
new file mode 100644
index 00000000..a9feba87
--- /dev/null
+++ b/imx8m/phyboard_polis/SharedBoardConfig.mk
@@ -0,0 +1,191 @@
+# -------@block_kernel_bootimg-------
+
+KERNEL_NAME := Image.lz4
+TARGET_KERNEL_ARCH := arm64
+
+LOADABLE_KERNEL_MODULE ?= true
+
+# CONFIG_VIDEO_MXC_CSI_CAMERA: mx6s_capture.ko, it's csi adapt driver which is the input of v4l2 framework
+# CONFIG_MXC_CAMERA_OV5640_MIPI_V2: ov5640_camera_mipi_v2.ko, sensor ov5640 driver, the input of mipi
+# CONFIG_MXC_MIPI_CSI: mxc_mipi_csi.ko, mipi driver which get the sensor data and send data to csi
+# CONFIG_VIDEO_MXC_CAPTURE: only enable compile dir drivers/media/platform/mxc/capture/, no need mxc_v4l2_capture.c which used in imx6/7
+# linux bsp team will add new configure for mxc_v4l2_capture.c in lf-5.4 branch.
+# CONFIG_SND_SOC_WM8524: snd-soc-wm8524.ko, wm8524 audio codec
+# CONFIG_SND_IMX_SOC: SoC Audio for Freescale i.MX CPUs
+# CONFIG_SND_SIMPLE_CARD: snd-soc-simple-card.ko snd-soc-simple-card-utils.ko, connect cpu and codec
+# CONFIG_SND_SOC_FSL_SAI: snd-soc-fsl-sai.ko, audio cpu, privide i2s
+# CONFIG_IMX_SDMA: imx-sdma.ko, sdma used for audio
+# CONFIG_SND_SOC_FSL_MICFIL: snd-soc-fsl-micfil.ko, used in audio mic
+# CONFIG_SND_SOC_IMX_PCM_DMA: imx-pcm-dma-common.ko, used in fsl_micfil
+# CONFIG_MXC_HANTRO_845: hantrodec_845s.ko vpu decodder
+# CONFIG_MXC_HANTRO_845_H1: hx280enc.ko vpu encoder
+# CONFIG_MXC_HANTRO_V4L2: vsiv4l2.ko  vpu v4l2 driver
+# CONFIG_RTC_DRV_SNVS: rtc-snvs.ko, snvs driver
+# CONFIG_FEC: fec.ko which depend on pps_core.ko and ptp.ko
+# CONFIG_AT803X_PHY: ethernet phy driver at803x.ko
+
+ifeq ($(LOADABLE_KERNEL_MODULE),true)
+BOARD_VENDOR_KERNEL_MODULES += \
+    $(KERNEL_OUT)/mm/zsmalloc.ko \
+    $(KERNEL_OUT)/drivers/block/zram/zram.ko \
+    $(KERNEL_OUT)/net/rfkill/rfkill.ko \
+    $(KERNEL_OUT)/net/wireless/cfg80211.ko \
+    $(KERNEL_OUT)/lib/crypto/libarc4.ko \
+    $(KERNEL_OUT)/net/mac80211/mac80211.ko \
+    $(KERNEL_OUT)/drivers/mxc/gpu-viv/galcore.ko \
+    $(KERNEL_OUT)/drivers/thermal/imx8mm_thermal.ko \
+    $(KERNEL_OUT)/drivers/media/platform/mxc/capture/mx6s_capture.ko \
+    $(KERNEL_OUT)/drivers/media/platform/mxc/capture/mxc_mipi_csi.ko \
+    $(KERNEL_OUT)/drivers/media/platform/mxc/capture/ov5640_camera_mipi_v2.ko \
+    $(KERNEL_OUT)/drivers/leds/leds-gpio.ko \
+    $(KERNEL_OUT)/drivers/leds/leds-pca995x.ko \
+    $(KERNEL_OUT)/drivers/gpio/gpio-pca953x.ko \
+    $(KERNEL_OUT)/sound/soc/fsl/imx-pcm-dma.ko \
+    $(KERNEL_OUT)/sound/soc/fsl/snd-soc-fsl-utils.ko \
+    $(KERNEL_OUT)/sound/soc/fsl/snd-soc-fsl-micfil.ko \
+    $(KERNEL_OUT)/sound/soc/fsl/snd-soc-fsl-asrc.ko \
+    $(KERNEL_OUT)/sound/soc/fsl/snd-soc-fsl-easrc.ko \
+    $(KERNEL_OUT)/sound/soc/fsl/snd-soc-fsl-sai.ko \
+    $(KERNEL_OUT)/sound/soc/fsl/snd-soc-fsl-spdif.ko \
+    $(KERNEL_OUT)/sound/soc/fsl/snd-soc-imx-spdif.ko \
+    $(KERNEL_OUT)/sound/soc/codecs/snd-soc-wm8524.ko \
+    $(KERNEL_OUT)/sound/soc/codecs/snd-soc-ak4458.ko \
+    $(KERNEL_OUT)/sound/soc/codecs/snd-soc-ak5558.ko \
+    $(KERNEL_OUT)/sound/soc/codecs/snd-soc-bt-sco.ko \
+    $(KERNEL_OUT)/sound/soc/generic/snd-soc-simple-card.ko \
+    $(KERNEL_OUT)/sound/soc/generic/snd-soc-simple-card-utils.ko \
+    $(KERNEL_OUT)/drivers/rpmsg/rpmsg_ns.ko \
+    $(KERNEL_OUT)/drivers/rpmsg/virtio_rpmsg_bus.ko \
+    $(KERNEL_OUT)/drivers/remoteproc/imx_rproc.ko \
+    $(KERNEL_OUT)/drivers/i2c/busses/i2c-rpmsg-imx.ko \
+    $(KERNEL_OUT)/sound/soc/fsl/imx-pcm-rpmsg.ko \
+    $(KERNEL_OUT)/sound/soc/fsl/snd-soc-fsl-rpmsg.ko \
+    $(KERNEL_OUT)/sound/soc/fsl/imx-audio-rpmsg.ko \
+    $(KERNEL_OUT)/sound/soc/codecs/snd-soc-rpmsg-ak4497.ko \
+    $(KERNEL_OUT)/sound/soc/fsl/snd-soc-imx-rpmsg.ko \
+    $(KERNEL_OUT)/sound/soc/fsl/snd-soc-imx-card.ko \
+    $(KERNEL_OUT)/drivers/mxc/hantro_845/hantrodec_845s.ko \
+    $(KERNEL_OUT)/drivers/mxc/hantro_845_h1/hx280enc.ko \
+    $(KERNEL_OUT)/drivers/mxc/hantro_v4l2/vsiv4l2.ko \
+    $(KERNEL_OUT)/drivers/rtc/rtc-snvs.ko \
+    $(KERNEL_OUT)/drivers/net/phy/at803x.ko \
+    $(KERNEL_OUT)/drivers/net/phy/realtek.ko \
+    $(KERNEL_OUT)/drivers/pps/pps_core.ko \
+    $(KERNEL_OUT)/drivers/ptp/ptp.ko \
+    $(KERNEL_OUT)/drivers/net/ethernet/freescale/fec.ko
+endif
+
+# CONFIG_TOUCHSCREEN_GOODIX: goodix_ts.ko, rm67199 mipi-panel touch driver module
+# CONFIG_TOUCHSCREEN_SYNAPTICS_DSX_I2C: synaptics_dsx_i2c.ko, rm67191 mipi-panel touch driver module
+# CONFIG_ZRAM: zram.ko compressed ram using LZ coding.
+# CONFIG_ZSMALLOC: zsmalloc.ko
+# CONFIG_CLK_IMX8MM: clk-imx8mm.ko
+# CONFIG_IMX8M_PM_DOMAINS: imx8m_pm_domains.ko, this driver still not upstream
+# CONFIG_IMX_GPCV2_PM_DOMAINS: gpcv2.ko, gpcv2-imx.ko
+# CONFIG_TIMER_IMX_SYS_CTR: timer-imx-sysctr.ko
+# CONFIG_PINCTRL_IMX8MM: pinctrl-imx8mm.ko
+# CONFIG_SERIAL_IMX: imx.ko
+# CONFIG_IMX2_WDT: imx2_wdt.ko
+# CONFIG_MFD_ROHM_BD718XX: rohm-bd718x7.ko
+# CONFIG_GPIO_MXC: gpio-generic.ko gpio-mxc.ko
+# CONFIG_MMC_SDHCI_ESDHC_IMX: sdhci-esdhc-imx.ko cqhci.ko
+# CONFIG_I2C_IMX:i2c-imx.ko
+# CONFIG_DMABUF_HEAPS_SYSTEM: system_heap.ko
+# CONFIG_DMABUF_HEAPS_CMA: cma_heap.ko
+# CONFIG_DMABUF_IMX: dma-buf-imx.ko
+# depend on clk module: reset-dispmix.ko, it will been select as m if clk build as m.
+# CONFIG_KEYBOARD_SNVS_PWRKEY: snvs_pwrkey.ko, snvs power key driver
+# CONFIG_IMX8MM_LCDIF_CORE: imx8mm-lcdif-core.ko
+# CONFIG_DRM_IMX: imxdrm.ko imx8mm-lcdif-crtc.ko
+# CONFIG_DRM_SEC_MIPI_DSIM: sec-dsim.ko
+# CONFIG_DRM_IMX_SEC_DSIM: sec_mipi_dsim-imx.ko
+# CONFIG_DRM_I2C_ADV7511: adv7511.ko
+# CONFIG_USB_CHIPIDEA_OF: usbmisc_imx.ko ci_hdrc_imx.ko
+# CONFIG_USB_CHIPIDEA: ci_hdrc.ko
+# CONFIG_NOP_USB_XCEIV: phy-generic.ko
+# CONFIG_TYPEC_TCPCI: tcpci.ko
+# CONFIG_USB_EHCI_HCD: ehci-hcd.ko
+# CONFIG_CFG80211: cfg80211.ko, cfg80211 - wireless configuration API
+# CONFIG_MAC80211: mac80211.ko, Generic IEEE 802.11 Networking Stack
+
+ifeq ($(LOADABLE_KERNEL_MODULE),true)
+BOARD_VENDOR_RAMDISK_KERNEL_MODULES +=     \
+    $(KERNEL_OUT)/drivers/clk/imx/mxc-clk.ko \
+    $(KERNEL_OUT)/drivers/clk/imx/clk-imx8mm.ko \
+    $(KERNEL_OUT)/drivers/soc/imx/soc-imx8m.ko \
+    $(KERNEL_OUT)/drivers/pmdomain/imx/imx8m-blk-ctrl.ko \
+    $(KERNEL_OUT)/drivers/soc/imx/imx8m_pm_domains.ko \
+    $(KERNEL_OUT)/drivers/pmdomain/imx/gpcv2.ko \
+    $(KERNEL_OUT)/drivers/pmdomain/imx/gpcv2-imx.ko \
+    $(KERNEL_OUT)/drivers/clocksource/timer-imx-sysctr.ko \
+    $(KERNEL_OUT)/drivers/soc/imx/busfreq-imx8mq.ko \
+    $(KERNEL_OUT)/drivers/pinctrl/freescale/pinctrl-imx.ko \
+    $(KERNEL_OUT)/drivers/pinctrl/freescale/pinctrl-imx8mm.ko \
+    $(KERNEL_OUT)/drivers/tty/serial/imx.ko \
+    $(KERNEL_OUT)/drivers/watchdog/imx2_wdt.ko \
+    $(KERNEL_OUT)/drivers/regulator/pca9450-regulator.ko \
+    $(KERNEL_OUT)/drivers/gpio/gpio-mxc.ko \
+    $(KERNEL_OUT)/drivers/perf/fsl_imx8_ddr_perf.ko \
+    $(KERNEL_OUT)/drivers/mfd/rohm-bd718x7.ko \
+    $(KERNEL_OUT)/drivers/regulator/rohm-regulator.ko \
+    $(KERNEL_OUT)/drivers/regulator/bd718x7-regulator.ko \
+    $(KERNEL_OUT)/drivers/cpufreq/cpufreq-dt.ko \
+    $(KERNEL_OUT)/drivers/cpufreq/imx-cpufreq-dt.ko \
+    $(KERNEL_OUT)/drivers/nvmem/nvmem-imx-ocotp.ko \
+    $(KERNEL_OUT)/drivers/mmc/host/sdhci-esdhc-imx.ko \
+    $(KERNEL_OUT)/drivers/mmc/host/cqhci.ko \
+    $(KERNEL_OUT)/drivers/i2c/busses/i2c-imx.ko \
+    $(KERNEL_OUT)/drivers/i2c/i2c-dev.ko \
+    $(KERNEL_OUT)/drivers/spi/spidev.ko \
+    $(KERNEL_OUT)/drivers/spi/spi-bitbang.ko \
+    $(KERNEL_OUT)/drivers/spi/spi-nxp-fspi.ko \
+    $(KERNEL_OUT)/drivers/spi/spi-imx.ko \
+    $(KERNEL_OUT)/lib/stmp_device.ko \
+    $(KERNEL_OUT)/drivers/dma/mxs-dma.ko \
+    $(KERNEL_OUT)/drivers/mmc/core/pwrseq_simple.ko \
+    $(KERNEL_OUT)/drivers/mailbox/imx-mailbox.ko \
+    $(KERNEL_OUT)/drivers/dma-buf/heaps/system_heap.ko \
+    $(KERNEL_OUT)/drivers/dma-buf/heaps/cma_heap.ko \
+    $(KERNEL_OUT)/drivers/dma-buf/dma-buf-imx.ko \
+    $(KERNEL_OUT)/drivers/input/keyboard/snvs_pwrkey.ko \
+    $(KERNEL_OUT)/drivers/input/touchscreen/goodix_ts.ko \
+    $(KERNEL_OUT)/drivers/input/touchscreen/synaptics_dsx/synaptics_dsx_i2c.ko \
+    $(KERNEL_OUT)/drivers/reset/reset-dispmix.ko \
+    $(KERNEL_OUT)/drivers/reset/reset-imx7.ko \
+    $(KERNEL_OUT)/drivers/gpu/imx/lcdif/imx8mm-lcdif-core.ko \
+    $(KERNEL_OUT)/drivers/gpu/drm/drm_dma_helper.ko \
+    $(KERNEL_OUT)/drivers/gpu/drm/bridge/adv7511/adv7511.ko \
+    $(KERNEL_OUT)/drivers/gpu/drm/bridge/sec-dsim.ko \
+    $(KERNEL_OUT)/drivers/gpu/drm/imx/imxdrm.ko \
+    $(KERNEL_OUT)/drivers/gpu/drm/imx/lcdif/imx8mm-lcdif-crtc.ko \
+    $(KERNEL_OUT)/drivers/gpu/drm/panel/panel-raydium-rm67191.ko \
+    $(KERNEL_OUT)/drivers/gpu/drm/imx/sec_mipi_dsim-imx.ko \
+    $(KERNEL_OUT)/drivers/usb/chipidea/usbmisc_imx.ko \
+    $(KERNEL_OUT)/drivers/usb/common/ulpi.ko \
+    $(KERNEL_OUT)/drivers/usb/chipidea/ci_hdrc_imx.ko \
+    $(KERNEL_OUT)/drivers/usb/chipidea/ci_hdrc.ko \
+    $(KERNEL_OUT)/drivers/usb/phy/phy-generic.ko \
+    $(KERNEL_OUT)/drivers/power/supply/dummy_battery.ko \
+    $(KERNEL_OUT)/drivers/dma/imx-sdma.ko \
+    $(KERNEL_OUT)/drivers/trusty/trusty-ipc.ko \
+    $(KERNEL_OUT)/drivers/trusty/trusty-core.ko \
+    $(KERNEL_OUT)/drivers/trusty/trusty-log.ko \
+    $(KERNEL_OUT)/drivers/trusty/trusty-virtio.ko
+else
+BOARD_VENDOR_RAMDISK_KERNEL_MODULES +=     \
+    $(KERNEL_OUT)/drivers/input/touchscreen/goodix_ts.ko \
+    $(KERNEL_OUT)/drivers/input/touchscreen/synaptics_dsx/synaptics_dsx_i2c.ko
+endif
+
+
+# -------@block_memory-------
+#Enable this to config 1GB ddr on evk_imx8mm
+LOW_MEMORY := false
+
+# -------@block_security-------
+#Enable this to include trusty support
+PRODUCT_IMX_TRUSTY := true
+
+# -------@block_storage-------
+# the bootloader image used in dual-bootloader OTA
+BOARD_OTA_BOOTLOADERIMAGE := u-boot-imx8mm.imx
diff --git a/imx8m/phyboard_polis/UbootKernelBoardConfig.mk b/imx8m/phyboard_polis/UbootKernelBoardConfig.mk
new file mode 100644
index 00000000..5cb3c49c
--- /dev/null
+++ b/imx8m/phyboard_polis/UbootKernelBoardConfig.mk
@@ -0,0 +1,21 @@
+TARGET_BOOTLOADER_POSTFIX := bin
+UBOOT_POST_PROCESS := true
+
+# u-boot target for phyboard_polis with LPDDR4 on board
+TARGET_BOOTLOADER_CONFIG := imx8mm:phycore-imx8mm_android_defconfig
+TARGET_BOOTLOADER_CONFIG += imx8mm-uuu:phycore-imx8mm_android_uuu_defconfig
+
+# imx8mm kernel defconfig
+TARGET_KERNEL_DEFCONFIG := imx8_phytec_defconfig
+ifeq ($(LOADABLE_KERNEL_MODULE),true)
+TARGET_KERNEL_GKI_DEFCONF:= imx8mm_gki.fragment
+else
+TARGET_KERNEL_GKI_DEFCONF := imx_v8_android_defconfig
+endif
+TARGET_KERNEL_ADDITION_DEFCONF := android_addition_defconfig
+
+# absolute path is used, not the same as relative path used in AOSP make
+TARGET_DEVICE_DIR := $(patsubst %/, %, $(dir $(realpath $(lastword $(MAKEFILE_LIST)))))
+
+# define bootloader rollback index
+BOOTLOADER_RBINDEX ?= 0
\ No newline at end of file
diff --git a/imx8m/phyboard_polis/phyboard_polis.mk b/imx8m/phyboard_polis/phyboard_polis.mk
new file mode 100644
index 00000000..094d0ee8
--- /dev/null
+++ b/imx8m/phyboard_polis/phyboard_polis.mk
@@ -0,0 +1,520 @@
+# -------@block_infrastructure-------
+
+CONFIG_REPO_PATH := device/nxp
+CURRENT_FILE_PATH :=  $(lastword $(MAKEFILE_LIST))
+IMX_DEVICE_PATH := $(strip $(patsubst %/, %, $(dir $(CURRENT_FILE_PATH))))
+
+PRODUCT_ENFORCE_ARTIFACT_PATH_REQUIREMENTS := true
+#Enable this to choose 32 bit user space build
+IMX_BUILD_32BIT_ROOTFS ?= false
+
+# configs shared between uboot, kernel and Android rootfs
+include $(IMX_DEVICE_PATH)/SharedBoardConfig.mk
+
+-include $(CONFIG_REPO_PATH)/common/imx_path/ImxPathConfig.mk
+include $(CONFIG_REPO_PATH)/imx8m/ProductConfigCommon.mk
+
+# -------@block_common_config-------
+
+# Overrides
+PRODUCT_NAME := phyboard_polis
+PRODUCT_DEVICE := phyboard_polis
+PRODUCT_MODEL := PHYBOARD_POLIS
+
+TARGET_BOOTLOADER_BOARD_NAME := phyboard_polis
+
+PRODUCT_CHARACTERISTICS := tablet
+
+DEVICE_PACKAGE_OVERLAYS := $(IMX_DEVICE_PATH)/overlay
+
+PRODUCT_COMPATIBLE_PROPERTY_OVERRIDE := true
+
+PRODUCT_VENDOR_PROPERTIES += ro.soc.manufacturer=nxp
+PRODUCT_VENDOR_PROPERTIES += ro.soc.model=IMX8MM
+PRODUCT_VENDOR_PROPERTIES += ro.crypto.metadata_init_delete_all_keys.enabled=true
+# -------@block_treble-------
+PRODUCT_FULL_TREBLE_OVERRIDE := true
+
+# -------@block_power-------
+PRODUCT_SOONG_NAMESPACES += vendor/nxp-opensource/imx/power
+PRODUCT_SOONG_NAMESPACES += hardware/google/pixel
+
+PRODUCT_COPY_FILES += \
+    $(IMX_DEVICE_PATH)/powerhint_imx8mm.json:$(TARGET_COPY_OUT_VENDOR)/etc/configs/powerhint_imx8mm.json
+
+# Do not skip charger_not_need trigger by default
+PRODUCT_DEFAULT_PROPERTY_OVERRIDES += \
+    vendor.skip.charger_not_need=0
+
+PRODUCT_PACKAGES += \
+    android.hardware.power-service.imx
+
+TARGET_VENDOR_PROP := $(LOCAL_PATH)/product.prop
+
+# Thermal HAL
+PRODUCT_PACKAGES += \
+    android.hardware.thermal-service.imx
+
+PRODUCT_COPY_FILES += \
+    $(IMX_DEVICE_PATH)/thermal_info_config_imx8mm.json:$(TARGET_COPY_OUT_VENDOR)/etc/configs/thermal_info_config_imx8mm.json
+
+
+# -------@block_app-------
+
+# Set permission for GMS packages
+PRODUCT_COPY_FILES += \
+    $(CONFIG_REPO_PATH)/imx8m/permissions/privapp-permissions-imx.xml:$(TARGET_COPY_OUT_PRODUCT)/etc/permissions/privapp.permissions-imx.xml
+
+PRODUCT_COPY_FILES += \
+    $(IMX_DEVICE_PATH)/app_whitelist.xml:system/etc/sysconfig/app_whitelist.xml
+
+# -------@block_kernel_bootimg-------
+
+# Enable this to support vendor boot and boot header v3, this would be a MUST for GKI
+TARGET_USE_VENDOR_BOOT ?= true
+
+# Allow LZ4 compression
+BOARD_RAMDISK_USE_LZ4 := true
+
+PRODUCT_PROPERTY_OVERRIDES += \
+    vendor.gki.enable=true
+BOARD_USES_GENERIC_KERNEL_IMAGE := true
+$(call inherit-product, $(SRC_TARGET_DIR)/product/generic_ramdisk.mk)
+
+# We load the fstab from device tree so this is not needed, but since no kernel modules are installed to vendor
+# boot ramdisk so far, we need this step to generate the vendor-ramdisk folder or build process would fail. This
+# can be deleted once we figure out what kernel modules should be put into the vendor boot ramdisk.
+ifeq ($(TARGET_USE_VENDOR_BOOT),true)
+PRODUCT_COPY_FILES += \
+    $(IMX_DEVICE_PATH)/fstab.nxp:$(TARGET_COPY_OUT_VENDOR_RAMDISK)/first_stage_ramdisk/fstab.nxp
+endif
+
+PRODUCT_COPY_FILES += \
+    $(IMX_DEVICE_PATH)/early.init.cfg:$(TARGET_COPY_OUT_VENDOR)/etc/early.init.cfg \
+    $(LINUX_FIRMWARE_IMX_PATH)/linux-firmware-imx/firmware/sdma/sdma-imx7d.bin:$(TARGET_COPY_OUT_VENDOR_RAMDISK)/lib/firmware/imx/sdma/sdma-imx7d.bin \
+    $(CONFIG_REPO_PATH)/common/init/init.insmod.sh:$(TARGET_COPY_OUT_VENDOR)/bin/init.insmod.sh \
+    $(IMX_DEVICE_PATH)/ueventd.nxp.rc:$(TARGET_COPY_OUT_VENDOR)/etc/ueventd.rc
+
+# -------@block_storage-------
+
+# support metadata checksum during first stage mount
+ifeq ($(TARGET_USE_VENDOR_BOOT),true)
+PRODUCT_PACKAGES += \
+    linker.vendor_ramdisk \
+    resizefs.vendor_ramdisk \
+    tune2fs.vendor_ramdisk
+endif
+
+#Enable this to use dynamic partitions for the readonly partitions not touched by bootloader
+TARGET_USE_DYNAMIC_PARTITIONS ?= true
+
+ifeq ($(TARGET_USE_DYNAMIC_PARTITIONS),true)
+  ifeq ($(TARGET_USE_VENDOR_BOOT),true)
+    $(call inherit-product, $(SRC_TARGET_DIR)/product/virtual_ab_ota/compression_with_xor.mk)
+  else
+    $(call inherit-product, $(SRC_TARGET_DIR)/product/virtual_ab_ota.mk)
+  endif
+  PRODUCT_USE_DYNAMIC_PARTITIONS := true
+  BOARD_BUILD_SUPER_IMAGE_BY_DEFAULT := true
+  BOARD_SUPER_IMAGE_IN_UPDATE_PACKAGE := true
+endif
+
+#Enable this to disable product partition build.
+IMX_NO_PRODUCT_PARTITION := false
+
+$(call inherit-product, $(SRC_TARGET_DIR)/product/emulated_storage.mk)
+
+
+PRODUCT_COPY_FILES += \
+    $(IMX_DEVICE_PATH)/fstab.nxp:$(TARGET_COPY_OUT_VENDOR)/etc/fstab.nxp
+
+TARGET_RECOVERY_FSTAB = $(IMX_DEVICE_PATH)/fstab.nxp
+
+ifneq ($(filter TRUE true 1,$(IMX_OTA_POSTINSTALL)),)
+  PRODUCT_PACKAGES += imx_ota_postinstall
+
+  AB_OTA_POSTINSTALL_CONFIG += \
+    RUN_POSTINSTALL_vendor=true \
+    POSTINSTALL_PATH_vendor=bin/imx_ota_postinstall \
+    FILESYSTEM_TYPE_vendor=erofs \
+    POSTINSTALL_OPTIONAL_vendor=false
+
+  PRODUCT_COPY_FILES += \
+    $(OUT_DIR)/target/product/$(firstword $(PRODUCT_DEVICE))/obj/UBOOT_COLLECTION/spl-imx8mm-trusty-dual.bin:$(TARGET_COPY_OUT_VENDOR)/etc/bootloader0.img
+  ifeq ($(BUILD_ENCRYPTED_BOOT),true)
+    PRODUCT_COPY_FILES += \
+      $(OUT_DIR)/target/product/$(firstword $(PRODUCT_DEVICE))/obj/UBOOT_COLLECTION/bootloader-imx8mm-trusty-dual.img:$(TARGET_COPY_OUT_VENDOR)/etc/bootloader_ab.img
+  endif
+endif
+
+# fastboot_imx_flashall scripts, imx-sdcard-partition script uuu_imx_android_flash scripts
+PRODUCT_COPY_FILES += \
+    $(CONFIG_REPO_PATH)/common/tools/fastboot_imx_flashall.bat:fastboot_imx_flashall.bat \
+    $(CONFIG_REPO_PATH)/common/tools/fastboot_imx_flashall.sh:fastboot_imx_flashall.sh \
+    $(CONFIG_REPO_PATH)/common/tools/imx-sdcard-partition.sh:imx-sdcard-partition.sh \
+    $(CONFIG_REPO_PATH)/common/tools/uuu_imx_android_flash.bat:uuu_imx_android_flash.bat \
+    $(CONFIG_REPO_PATH)/common/tools/uuu_imx_android_flash.sh:uuu_imx_android_flash.sh
+
+# -------@block_security-------
+
+# Include keystore attestation keys and certificates.
+ifeq ($(PRODUCT_IMX_TRUSTY),true)
+-include $(IMX_SECURITY_PATH)/attestation/imx_attestation.mk
+endif
+
+
+ifeq ($(PRODUCT_IMX_TRUSTY),true)
+PRODUCT_COPY_FILES += \
+    $(CONFIG_REPO_PATH)/common/security/rpmb_key_test.bin:rpmb_key_test.bin \
+    $(CONFIG_REPO_PATH)/common/security/testkey_public_rsa4096.bin:testkey_public_rsa4096.bin
+endif
+
+# Keymaster HAL
+ifeq ($(PRODUCT_IMX_TRUSTY),true)
+PRODUCT_PACKAGES += \
+    android.hardware.security.keymint-service.rust.trusty
+endif
+
+PRODUCT_PACKAGES += \
+    android.hardware.security.keymint-service-imx
+
+# Confirmation UI
+ifeq ($(PRODUCT_IMX_TRUSTY),true)
+PRODUCT_PACKAGES += \
+    android.hardware.confirmationui-service.trusty
+endif
+
+# new gatekeeper HAL
+PRODUCT_PACKAGES += \
+    android.hardware.gatekeeper-service-imx
+
+# Add Trusty OS backed gatekeeper and secure storage proxy
+ifeq ($(PRODUCT_IMX_TRUSTY),true)
+PRODUCT_PACKAGES += \
+    android.hardware.gatekeeper-service.trusty \
+    storageproxyd \
+    imx_dek_extractor \
+    imx_dek_inserter
+endif
+
+# Add oem unlocking option in settings.
+PRODUCT_PROPERTY_OVERRIDES += ro.frp.pst=/dev/block/by-name/presistdata
+
+ifeq ($(PRODUCT_IMX_TRUSTY),true)
+#Oemlock HAL support
+PRODUCT_PACKAGES += \
+    android.hardware.oemlock-service.imx \
+    android.hardware.oemlock-service-software.imx
+endif
+
+# Secretkeeper HAL
+PRODUCT_PACKAGES += \
+    com.android.hardware.security.secretkeeper
+
+ifeq ($(PRODUCT_IMX_TRUSTY),true)
+PRODUCT_PACKAGES += \
+    android.hardware.security.secretkeeper.trusty
+endif
+
+# Specify rollback index for boot and vbmeta partition
+ifneq ($(AVB_RBINDEX),)
+BOARD_AVB_ROLLBACK_INDEX := $(AVB_RBINDEX)
+else
+BOARD_AVB_ROLLBACK_INDEX := 0
+endif
+
+ifneq ($(AVB_BOOT_RBINDEX),)
+BOARD_AVB_BOOT_ROLLBACK_INDEX := $(AVB_BOOT_RBINDEX)
+else
+BOARD_AVB_BOOT_ROLLBACK_INDEX := 0
+endif
+
+ifneq ($(AVB_INIT_BOOT_RBINDEX),)
+BOARD_AVB_INIT_BOOT_ROLLBACK_INDEX := $(AVB_INIT_BOOT_RBINDEX)
+else
+BOARD_AVB_INIT_BOOT_ROLLBACK_INDEX := 0
+endif
+
+$(call  inherit-product-if-exists, vendor/nxp-private/security/nxp_security.mk)
+
+# Resume on Reboot support
+PRODUCT_PACKAGES += \
+    android.hardware.rebootescrow-service.default
+
+PRODUCT_PROPERTY_OVERRIDES += \
+    ro.rebootescrow.device=/dev/block/pmem0
+
+#DRM Widevine 1.4 L3 support
+PRODUCT_PACKAGES += \
+    android.hardware.drm-service.clearkey \
+    libwvdrmcryptoplugin
+
+TARGET_BUILD_WIDEVINE :=
+TARGET_BUILD_WIDEVINE_USE_PREBUILT := true
+
+$(call inherit-product-if-exists, vendor/nxp-private/widevine/apex/device.mk)
+
+# -------@block_audio-------
+
+# Audio card json
+PRODUCT_COPY_FILES += \
+    $(CONFIG_REPO_PATH)/common/audio-json/wm8524_config.json:$(TARGET_COPY_OUT_VENDOR)/etc/configs/audio/wm8524_config.json \
+    $(CONFIG_REPO_PATH)/common/audio-json/spdif_config.json:$(TARGET_COPY_OUT_VENDOR)/etc/configs/audio/spdif_config.json \
+    $(CONFIG_REPO_PATH)/common/audio-json/micfil_config.json:$(TARGET_COPY_OUT_VENDOR)/etc/configs/audio/micfil_config.json \
+    $(CONFIG_REPO_PATH)/common/audio-json/ak4458_config.json:$(TARGET_COPY_OUT_VENDOR)/etc/configs/audio/ak4458_config.json \
+    $(CONFIG_REPO_PATH)/common/audio-json/ak4497_config.json:$(TARGET_COPY_OUT_VENDOR)/etc/configs/audio/ak4497_config.json \
+    $(CONFIG_REPO_PATH)/common/audio-json/btsco_config.json:$(TARGET_COPY_OUT_VENDOR)/etc/configs/audio/btsco_config.json \
+    $(CONFIG_REPO_PATH)/common/audio-json/readme.txt:$(TARGET_COPY_OUT_VENDOR)/etc/configs/audio/readme.txt
+
+PRODUCT_COPY_FILES += \
+    $(FSL_PROPRIETARY_PATH)/fsl-proprietary/mcu-sdk/imx8mm/imx8mm_mcu_demo.img:imx8mm_mcu_demo.img \
+    $(IMX_DEVICE_PATH)/audio_effects.xml:$(TARGET_COPY_OUT_VENDOR)/etc/audio_effects.xml \
+    $(IMX_DEVICE_PATH)/audio_policy_configuration.xml:$(TARGET_COPY_OUT_VENDOR)/etc/audio_policy_configuration.xml \
+    $(IMX_DEVICE_PATH)/usb_audio_policy_configuration-direct-output.xml:$(TARGET_COPY_OUT_VENDOR)/etc/usb_audio_policy_configuration-direct-output.xml
+
+
+# -------@block_camera-------
+PRODUCT_COPY_FILES += \
+    $(IMX_DEVICE_PATH)/camera_config_imx8mm.json:$(TARGET_COPY_OUT_VENDOR)/etc/configs/camera_config_imx8mm.json \
+    $(IMX_DEVICE_PATH)/external_camera_config.xml:$(TARGET_COPY_OUT_VENDOR)/etc/external_camera_config.xml
+
+PRODUCT_SOONG_NAMESPACES += hardware/google/camera
+PRODUCT_SOONG_NAMESPACES += vendor/nxp-opensource/imx/camera
+
+# Add WebCam option in settings
+PRODUCT_VENDOR_PROPERTIES += ro.usb.uvc.enabled=true
+
+# -------@block_display-------
+
+PRODUCT_AAPT_CONFIG += xlarge large tvdpi hdpi xhdpi xxhdpi
+
+# -----some limiataion of overlay/g2d in hwcomposer3 --------
+SOONG_CONFIG_NAMESPACES += nxp_hwc
+SOONG_CONFIG_nxp_hwc += g2d_ip
+SOONG_CONFIG_nxp_hwc_g2d_ip := VIV
+
+# HWC3 HAL
+PRODUCT_PACKAGES += \
+    android.hardware.graphics.composer3-service.imx
+
+# define frame buffer count
+PRODUCT_DEFAULT_PROPERTY_OVERRIDES += \
+    ro.surface_flinger.max_frame_buffer_acquired_buffers=3
+
+# set game default frame rate override
+PRODUCT_DEFAULT_PROPERTY_OVERRIDES += \
+    ro.surface_flinger.game_default_frame_rate_override=60
+
+# Gralloc HAL
+PRODUCT_PACKAGES += \
+    android.hardware.graphics.allocator-service.imx \
+    mapper.imx
+
+# 2d test
+ifneq (,$(filter userdebug eng, $(TARGET_BUILD_VARIANT)))
+PRODUCT_PACKAGES += 2d-test
+endif
+
+# -------@block_gpu-------
+
+PRODUCT_PACKAGES += \
+    libEGL_VIVANTE \
+    libGLESv1_CM_VIVANTE \
+    libGLESv2_VIVANTE \
+    gralloc_viv.$(TARGET_BOARD_PLATFORM) \
+    libGAL \
+    libGLSLC \
+    libVSC \
+    libCLC \
+    libg2d-viv \
+    libgpuhelper
+
+PRODUCT_VENDOR_PROPERTIES += \
+    ro.hardware.egl = VIVANTE
+
+# -------@block_wifi-------
+
+#LPDDR4 board, NXP wifi supplicant overlay
+PRODUCT_COPY_FILES += \
+    $(IMX_DEVICE_PATH)/init.imx8mm.lpddr4.rc:$(TARGET_COPY_OUT_VENDOR)/etc/init/hw/init.nxp.additional.rc \
+    $(CONFIG_REPO_PATH)/common/wifi/wpa_supplicant_overlay.conf:$(TARGET_COPY_OUT_VENDOR)/etc/wifi/wpa_supplicant_overlay.conf
+
+PRODUCT_COPY_FILES += \
+    $(CONFIG_REPO_PATH)/common/wifi/p2p_supplicant_overlay.conf:$(TARGET_COPY_OUT_VENDOR)/etc/wifi/p2p_supplicant_overlay.conf \
+
+# WiFi HAL
+PRODUCT_PACKAGES += \
+    android.hardware.wifi-service \
+    wificond
+
+# WiFi RRO
+PRODUCT_PACKAGES += \
+    WifiOverlay
+
+# NXP 8987 WiFi Firmware
+PRODUCT_COPY_FILES += \
+    vendor/nxp/imx-firmware/nxp/FwImage_8987/sduart8987_combo.bin:vendor/firmware/sduart8987_combo.bin \
+    vendor/nxp/imx-firmware/nxp/FwImage_IW612_SD/sduart_nw61x_v1.bin.se:vendor/firmware/sduart_nw61x_v1.bin.se \
+    vendor/nxp/imx-firmware/nxp/android_wifi_mod_para.conf:vendor/firmware/wifi_mod_para_sd8987.conf
+
+# Wifi regulatory
+PRODUCT_COPY_FILES += \
+    external/wireless-regdb/regulatory.db:$(TARGET_COPY_OUT_VENDOR)/firmware/regulatory.db \
+    external/wireless-regdb/regulatory.db.p7s:$(TARGET_COPY_OUT_VENDOR)/firmware/regulatory.db.p7s
+
+# -------@block_bluetooth-------
+
+# Bluetooth HAL
+PRODUCT_PACKAGES += \
+    android.hardware.bluetooth \
+    android.hardware.bluetooth-service.default.nxp
+
+# NXP 8987 Bluetooth vendor config
+PRODUCT_PACKAGES += \
+    bt_vendor.conf
+
+# Bluetooth LE Audio
+PRODUCT_PRODUCT_PROPERTIES += \
+    ro.bluetooth.leaudio_offload.supported=false \
+    persist.bluetooth.leaudio_offload.disabled=false \
+    ro.bluetooth.leaudio_switcher.supported=true
+
+# -------@block_usb-------
+
+# Usb HAL
+PRODUCT_PACKAGES += \
+    android.hardware.usb-service.imx \
+    android.hardware.usb.gadget-service.imx
+
+PRODUCT_COPY_FILES += \
+    $(IMX_DEVICE_PATH)/init.usb.rc:$(TARGET_COPY_OUT_VENDOR)/etc/init/hw/init.nxp.usb.rc
+
+# -------@block_multimedia_codec-------
+
+# Vendor seccomp policy files for media components:
+PRODUCT_COPY_FILES += \
+    $(IMX_DEVICE_PATH)/seccomp/mediacodec-seccomp.policy:vendor/etc/seccomp_policy/mediacodec.policy \
+    $(IMX_DEVICE_PATH)/seccomp/mediaextractor-seccomp.policy:vendor/etc/seccomp_policy/mediaextractor.policy
+
+
+PRODUCT_PACKAGES += \
+    libg1 \
+    libhantro \
+    libcodec \
+    libhantro_h1 \
+    libcodec_enc \
+    DirectAudioPlayer
+
+# imx c2 codec binary
+PRODUCT_PACKAGES += \
+    lib_vpu_wrapper \
+    lib_imx_c2_videodec \
+    lib_imx_c2_videodec_common \
+    lib_imx_c2_videoenc_common \
+    lib_imx_c2_videoenc \
+    lib_imx_c2_v4l2_dev \
+    lib_imx_c2_v4l2_dec \
+    lib_imx_c2_v4l2_enc \
+    lib_imx_c2_unia_pre_filter \
+    lib_imx_c2_filter_device_factory \
+    lib_imx_c2_filter_device_g2d \
+    libc2filterplugin \
+    c2_component_register \
+    c2_component_register_ms \
+    c2_component_register_ra
+
+ifeq ($(PREBUILT_FSL_IMX_CODEC),true)
+ifneq ($(IMX_BUILD_32BIT_ROOTFS),true)
+INSTALL_64BIT_LIBRARY := true
+endif
+endif
+
+# -------@block_memory-------
+
+# Include Android Go config for low memory device.
+ifeq ($(LOW_MEMORY),true)
+  $(call inherit-product, build/target/product/go_defaults.mk)
+endif
+
+# -------@block_miscellaneous-------
+
+# Copy device related config and binary to board
+PRODUCT_COPY_FILES += \
+    $(IMX_DEVICE_PATH)/init.imx8mm.rc:$(TARGET_COPY_OUT_VENDOR)/etc/init/hw/init.nxp.imx8mm.rc \
+    $(IMX_DEVICE_PATH)/init.rc:$(TARGET_COPY_OUT_VENDOR)/etc/init/hw/init.nxp.rc \
+    $(IMX_DEVICE_PATH)/required_hardware.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/required_hardware.xml
+
+ifeq ($(TARGET_USE_VENDOR_BOOT),true)
+  PRODUCT_COPY_FILES += \
+    $(IMX_DEVICE_PATH)/init.recovery.nxp.rc:$(TARGET_COPY_OUT_VENDOR_RAMDISK)/init.recovery.nxp.rc
+else
+  PRODUCT_COPY_FILES += \
+    $(IMX_DEVICE_PATH)/init.recovery.nxp.rc:root/init.recovery.nxp.rc
+endif
+
+# Display Device Config
+PRODUCT_COPY_FILES += \
+    device/nxp/imx8m/displayconfig/display_port_1.xml:$(TARGET_COPY_OUT_VENDOR)/etc/displayconfig/display_port_1.xml
+
+# ONLY devices that meet the CDD's requirements may declare these features
+PRODUCT_COPY_FILES += \
+    frameworks/native/data/etc/android.hardware.audio.output.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.audio.output.xml \
+    frameworks/native/data/etc/android.hardware.bluetooth_le.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.bluetooth_le.xml \
+    frameworks/native/data/etc/android.hardware.camera.front.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.camera.front.xml \
+    frameworks/native/data/etc/android.hardware.camera.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.camera.xml \
+    frameworks/native/data/etc/android.hardware.ethernet.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.ethernet.xml \
+    frameworks/native/data/etc/android.hardware.screen.landscape.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.screen.landscape.xml \
+    frameworks/native/data/etc/android.hardware.screen.portrait.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.screen.portrait.xml \
+    frameworks/native/data/etc/android.hardware.touchscreen.multitouch.distinct.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.touchscreen.multitouch.distinct.xml \
+    frameworks/native/data/etc/android.hardware.touchscreen.multitouch.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.touchscreen.multitouch.xml \
+    frameworks/native/data/etc/android.hardware.touchscreen.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.touchscreen.xml \
+    frameworks/native/data/etc/android.hardware.usb.accessory.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.usb.accessory.xml \
+    frameworks/native/data/etc/android.hardware.usb.host.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.usb.host.xml \
+    frameworks/native/data/etc/android.software.vulkan.deqp.level-2024-03-01.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.software.vulkan.deqp.level.xml \
+    frameworks/native/data/etc/android.software.opengles.deqp.level-2024-03-01.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.software.opengles.deqp.level.xml \
+    frameworks/native/data/etc/android.hardware.wifi.direct.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.wifi.direct.xml \
+    frameworks/native/data/etc/android.hardware.wifi.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.wifi.xml \
+    frameworks/native/data/etc/android.hardware.wifi.passpoint.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.wifi.passpoint.xml \
+    frameworks/native/data/etc/android.software.app_widgets.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.software.app_widgets.xml \
+    frameworks/native/data/etc/android.software.backup.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.software.backup.xml \
+    frameworks/native/data/etc/android.software.device_admin.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.software.device_admin.xml \
+    frameworks/native/data/etc/android.software.managed_users.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.software.managed_users.xml \
+    frameworks/native/data/etc/android.software.midi.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.software.midi.xml \
+    frameworks/native/data/etc/android.software.print.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.software.print.xml \
+    frameworks/native/data/etc/android.software.sip.voip.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.software.sip.voip.xml \
+    frameworks/native/data/etc/android.software.verified_boot.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.software.verified_boot.xml \
+    frameworks/native/data/etc/android.software.voice_recognizers.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.software.voice_recognizers.xml \
+    frameworks/native/data/etc/android.software.activities_on_secondary_displays.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.software.activities_on_secondary_displays.xml \
+    frameworks/native/data/etc/android.software.picture_in_picture.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.software.picture_in_picture.xml \
+    frameworks/native/data/etc/android.software.credentials.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.software.credentials.xml
+
+PERMISSION_EXTCAM ?= true
+ifeq ($(PERMISSION_EXTCAM),true)
+PRODUCT_COPY_FILES += \
+    frameworks/native/data/etc/android.hardware.camera.external.xml:vendor/etc/permissions/android.hardware.camera.external.xml
+endif
+
+# trusty loadable apps
+PRODUCT_COPY_FILES += \
+    vendor/nxp/fsl-proprietary/uboot-firmware/imx8m/confirmationui-imx8mm.app:/vendor/firmware/tee/confirmationui.app
+
+# Keymint configuration
+PRODUCT_COPY_FILES += \
+    frameworks/native/data/etc/android.software.device_id_attestation.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.software.device_id_attestation.xml
+
+# Add Virtualization support
+$(call inherit-product, packages/modules/Virtualization/apex/product_packages.mk)
+
+# Included GMS package
+ifeq ($(filter TRUE true 1,$(IMX_BUILD_32BIT_ROOTFS) $(IMX_BUILD_32BIT_64BIT_ROOTFS)),)
+$(call inherit-product-if-exists, vendor/partner_gms/products/gms_64bit_only.mk)
+else
+$(call inherit-product-if-exists, vendor/partner_gms/products/gms.mk)
+endif
+PRODUCT_SOONG_NAMESPACES += vendor/partner_gms
+
+PRODUCT_PACKAGES += \
+    privapp_whitelist_com.android.emergency
+
+# Add imx private apps
+$(call inherit-product-if-exists, vendor/nxp-private/imx-apps/imx-private-app.mk)
-- 
2.34.1

